File: 18/solution.cpp
class Solution {
public:
    // Try to formalize to 2sum
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        
        for (int i=0; i<nums.size(); i++) {
            int remainingTarget = target - nums[i];
            for (int j=i+1; j<nums.size(); j++) {
                int remainingSecondTarget = remainingTarget - nums[j];
                int left = j+1;
                int right = nums.size() - 1;
                
                // 2sum structure
                while (left < right) {
                    int sum = nums[left] + nums[right];
                    if (sum < remainingSecondTarget) left++;
                    else if (sum > remainingSecondTarget) right--;
                    else {
                        vector<int> ans(4, 0);
                        ans[0] = nums[i];
                        ans[1] = nums[j];
                        ans[2] = nums[left];
                        ans[3] = nums[right];
                        
                        result.push_back(ans);
                        
                        // Skip duplicated numbers
                        while(nums[left] == ans[2] && left < right) left++;
                        while(nums[right] == ans[3] && left < right) right--;
                    }
                }
                // Skip duplicated numbers
                while(j < nums.size() - 1 && nums[j] == nums[j+1]) j++;
            }
            // Skip duplicated numbers
            while(i < nums.size() - 1 && nums[i] == nums[i+1]) i++;
        }
        
        return result;
    }
};

File: 18/solution.go
import "sort"

func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	result := make([][]int, 0)

	for firstIdx := 0; firstIdx < len(nums); firstIdx++ {
		remainTarget := target - nums[firstIdx]

		for secondIdx := firstIdx + 1; secondIdx < len(nums); secondIdx++ {
			remainSecondTarget := remainTarget - nums[secondIdx]
			left := secondIdx + 1
			right := len(nums) - 1

			for left < right {
				sum := nums[left] + nums[right]

				if sum < remainSecondTarget {
					left++
				} else if sum > remainSecondTarget {
					right--
				} else {
					ans := make([]int, 4)
					ans[0] = nums[firstIdx]
					ans[1] = nums[secondIdx]
					ans[2] = nums[left]
					ans[3] = nums[right]

					result = append(result, ans)

					for left < right && nums[left] == ans[2] {
						left++
					}

					for left < right && nums[right] == ans[3] {
						right--
					}
				}
			}

			for secondIdx < len(nums)-1 && nums[secondIdx] == nums[secondIdx+1] {
				secondIdx++
			}
		}

		for firstIdx < len(nums)-1 && nums[firstIdx] == nums[firstIdx+1] {
			firstIdx++
		}
	}

	return result
}